### PR TITLE
Fix: Improve autoconfig logging (fixes #5119)

### LINKED ISSUE #5119: Improve autoconfig debug logging
It's difficult to troubleshoot autoconfig issues because of a lack of debug logging.  Additional debug statements should be inserted to log out the name of files being processed, the constructed options that are used when linting, and (importantly) the results registry from which config decisions are made. 

Additionally, parsing errors should be displayed to user regardless of debug mode.  Otherwise, if code is unparsable by Espree, it is impossible for the user to tell why autoconfig failed.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/30397451-improve-autoconfig-debug-logging?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by IanVS**
This also adds additional tests to bring the coverage back up, and fixes some bugs which were found during testing.

### CODE REVIEW

File: lib/util/source-code-util.js [hunk 1]
```diff
@@ -25,14 +29,19 @@ var assign = require("object-assign"),
...
     var opts = assign({}, options, { rules: {}});
     var cli = new CLIEngine(opts);
-    cli.executeOnFiles([filename]);
+    var results = cli.executeOnFiles([filename]);
+    if (results && results.results[0] && results.results[0].messages[0] && results.results[0].messages[0].fatal) {
+        return [null, results.results[0].messages[0].message];
+    }
     var sourceCode = eslint.getSourceCode();
-    return sourceCode;
+    return [sourceCode, null];
```
Comment (ilyavolodin) [inline lib/util/source-code-util.js]: Wouldn't it be easier to throw an error if you can't parse the file and run the code inside try/catch statement?

Comment (IanVS) [inline lib/util/source-code-util.js]: It __would__ be great to throw an error, but for the life of me I can't figure out how to get inquirer to handle or bubble up errors.  Instead, it just swallows them silently.  :-/

**Comment by gyandeeps** [comment]
@IanVS Can we plz use this PR to increase code coverage for this functionality? 
Initially with this functionality our coverage got drooped. It takes a big effort later on to pick up code coverage where a single person has to go all out on the whole code base (Which I did recently). Specially the branch coverage.

**Comment by IanVS** [comment]
Yes, I planned to add some more tests at some point anyway.  Is there a way to find the areas of code that are not covered?  Do I just need to go read the istanbul docs?

**Comment by gyandeeps** [comment]
You can open up index file inside eslint\coverage\lcov-report folder on your machine. That where istanbul creates the code coverage. Its an html file so it will give you all the information.

**Comment by ilyavolodin** [comment]
To be more specific, once you run `npm test` locally, istanbul will create `/coverage/` folder with all coverage information you might need. Just open it up in the browser and you should see full code coverage and can dig into specific files/functions.

File: lib/util/source-code-util.js (hunk 1 above)
Comment (nzakas) [inline lib/util/source-code-util.js]: You need to use a try-catch whjen you call this function and then pass the error to cb().

Comment (IanVS) [inline lib/util/source-code-util.js]: Cool, thanks.  I'll give that a try.

**Comment by IanVS** [comment]
OK, with the latest change the whole process will error if a parsing error is encountered (thanks to @nzakas explaining how to throw an error).  I think this is better than creating a config as long as at least __one__ file was parsable (which is what would have happened before).

I still need to add some more tests to increase coverage back up.

**Comment by IanVS** [comment]
I've brought the code coverage back up, and sure enough I found a few bugs. 

This also adds support for the new rules with schemas nested under a `meta` property.

I believe I've addressed all of the comments so far.

**Comment by gyandeeps** [comment]
Excellent work thr @IanVS ... Thanks a lot. :)

**Comment by nzakas** [comment]
@ilyavolodin want to double check the meta stuff?

**Comment by ilyavolodin** [comment]
LGTM

**Pull request merged**